In [26]:
import pandas as pd
import numpy as np
import math

train = pd.read_csv('crime-train.txt', delimiter='\t')
test = pd.read_csv('crime-test.txt', delimiter='\t')
train.head()
#test.head()

ViolentCrimesPerPop  population  householdsize  agePct12t21  agePct12t29  \
0                 0.67       -0.45          -1.85        -1.06         0.67   
1                 0.43       -0.45          -0.27        -0.22        -0.17   
2                 0.12       -0.14           1.87         0.55         0.04   
3                 0.03       -0.38           0.53        -0.28        -0.79   
4                 0.14       -0.30          -1.12        -0.74        -0.10   

   agePct16t24  agePct65up  numbUrban  pctUrban  medIncome  ...  NumStreet  \
0         0.08       -0.85      -0.34      0.68      -0.24  ...      -0.23   
1        -0.34       -0.58      -0.50     -1.57      -0.29  ...      -0.23   
2         0.02       -1.19      -0.03      0.68       1.05  ...      -0.23   
3        -0.64       -0.35      -0.34      0.46       0.66  ...      -0.23   
4        -0.40       -0.30      -0.19      0.68       0.76  ...      -0.23   

   PctForeignBorn  PctBornSameState  PctSameHouse85  PctSameCity85  \
0           -0.02             -0.53           -1.08          -0.13   
1           -0.33             -0.58            0.03           0.22   
2           -0.11             -1.51            1.07           0.07   
3           -0.46              0.54            0.58          -0.08   
4            2.10             -0.92           -0.25           0.52   

   PctSameState85  LandArea  PopDens  PctUsePubTrans  LemasPctOfficDrugUn  
0           -0.66     -0.41    -0.56            1.26                -0.39  
1           -0.46     -0.50    -0.11           -0.62                -0.39  
2           -0.01     -0.41     0.77            0.52                -0.39  
3           -0.61     -0.23    -0.70           -0.62                -0.39  
4           -0.06     -0.50     1.71           -0.27                -0.39  

[5 rows x 96 columns]

In [27]:
# store expected outcomes (1st column) in y_train
y_train = train['ViolentCrimesPerPop']
# store everything else in X_train
X_train = train.drop('ViolentCrimesPerPop',axis=1)

# we also have to convert values from string to float
y_train = np.float64(y_train)
X_train = np.float64(X_train)

# append 1s to the end of X_train
ones = np.ones(len(X_train))
X_train = np.column_stack((X_train, ones))

# turn array into a matrix (with one column)
y_train = np.reshape(y_train, (-1,1))

print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)

X_train shape: (1595, 96)
y_train shape: (1595, 1)


In [28]:
y_test = test['ViolentCrimesPerPop']
X_test = test.drop('ViolentCrimesPerPop',axis=1)

y_test = np.float64(y_test)
X_test = np.float64(X_test)

ones = np.ones(len(X_test))
X_test = np.column_stack((X_test, ones))

# turn the array into a matrix (with one column)
y_test = np.reshape(y_test, (-1,1))

print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_test shape: (399, 96)
y_test shape: (399, 1)


In [29]:
# same RMSE function as used in last lab
def RMSE(prediction, actual):
    n = len(prediction)
    diff = prediction - actual
    total = 0
    
    for instance in diff:
        total += (instance ** 2)
        
    total_error = math.sqrt(total/n)
    return total_error

In [30]:
# performs Ridge Regression and returns list of predicted outcomes that each correspond to their actual values
def ridge_regression(x, y, test, lambda_val):
    left = np.linalg.inv(np.dot(x.T, x) + (lambda_val * np.identity(len(x.T))))
    right = np.dot(x.T, y)
    w = np.dot(left, right)
    prediction = []
    for Xtest in test:
        prediction.append(np.dot(Xtest.T, w))
    return prediction


lambda_val = 400
RMSEs = np.zeros((10,5))
minimum = 0

# do k-fold cross validation with k = 5
for segment in range(len(RMSEs)):
    for k in range(5):
        # train with 4/5 of dataset
        X_train_k = np.concatenate((X_train[0:int(k*(len(X_train)/5))], X_train[int((k+1)*(len(X_train)/5)):int(len(X_train))]))
        y_train_k = np.concatenate((y_train[0:int(k*(len(y_train)/5))], y_train[int((k+1)*(len(y_train)/5)):int(len(y_train))]))
        
        # use remaining 1/5 dataset to validate results
        X_validate_k = X_train[int(k*(len(X_train)/5)):int((k+1)*(len(X_train)/5))]
        y_validate_k = y_train[int(k*(len(y_train)/5)):int((k+1)*(len(y_train)/5))]

        # perform ridge regression on each fold and each lambda 
        prediction = ridge_regression(X_train_k, y_train_k, X_validate_k, lambda_val)

        # for each lambda, put all 5 of the different RMSEs on the same row
        RMSEs[segment, k] = RMSE(prediction, y_validate_k)
    lambda_val /= 2
print("RMSE Values:\n", RMSEs, "\n")
    
# get average RMSE for each lambda and choose optimal lambda
for i in range(len(RMSEs)-1):
    if (sum(RMSEs[i+1]) / 5) < (sum(RMSEs[minimum]) / 5):
        minimum = i+1

for j in range(len(RMSEs)):
    print("𝜆 =", (400 / (2 ** j)), ":", sum(RMSEs[j]) / 5)

optimal_lambda = 400 / (2 ** minimum)

print("\nThe optimal 𝜆 is:", optimal_lambda)
print("Training RMSE for ridge regression using closed form solution is", (sum(RMSEs[minimum]) / 5))

RMSE Values:
 [[0.16031626 0.14530191 0.15755165 0.14669202 0.13770204]
 [0.1504947  0.13916403 0.14576515 0.13872299 0.12932324]
 [0.14641669 0.13761947 0.13961026 0.13589421 0.12684322]
 [0.14511564 0.13746282 0.1363316  0.13498737 0.1268823 ]
 [0.1449868  0.13758002 0.13455271 0.13464334 0.12781642]
 [0.14528919 0.13780086 0.13365482 0.13444721 0.12891979]
 [0.14570088 0.13815894 0.13329282 0.13430567 0.12988101]
 [0.14610098 0.13865628 0.13324682 0.13421881 0.1306319 ]
 [0.1464384  0.13923262 0.13336903 0.13420359 0.13119456]
 [0.14668497 0.13978827 0.13355288 0.13425675 0.13159847]] 

𝜆 = 400.0 : 0.14951277619075792
𝜆 = 200.0 : 0.14069401845317783
𝜆 = 100.0 : 0.1372767722318658
𝜆 = 50.0 : 0.13615594392755512
𝜆 = 25.0 : 0.13591585919851987
𝜆 = 12.5 : 0.13602237446230692
𝜆 = 6.25 : 0.13626786494076723
𝜆 = 3.125 : 0.13657095693642923
𝜆 = 1.5625 : 0.1368876407319747
𝜆 = 0.78125 : 0.13717626806867203

The optimal 𝜆 is: 25.0
Training RMSE for ridge regression using closed form solution 

In [31]:
# compute loss function for linear regression
def linear_gradient_descent(x, y, w):
    left = (y - np.dot(x, w)).T
    right = (y - np.dot(x, w))
    min_w = np.dot(left, right)
    return min_w

alpha = 5*(10**-5)

# use gradient descent until convergence to obtain w
L1 = 1
L0 = 0
# generate Gaussian distribution for initial value based off how many features we have
w0 = np.random.normal(0, 1, (X_train.shape[1],1))

while (abs(L1 - L0) > (10**-5)):
    w1 = w0 - (alpha * np.dot(X_train.T, (np.dot(X_train, w0) - y_train)))
    L0 = linear_gradient_descent(X_train, y_train, w0)
    L1 = linear_gradient_descent(X_train, y_train, w1)
    w0 = w1

In [32]:
#  takes in an NxK array and returns predictions in Nx1 array
def problem2(w_value, matrix):
    prediction = []
    for x in matrix:
        prediction.append(np.dot(w_value.T, x))
    return prediction

train_prediction = []
test_prediction = []

# make predictions and calculate RMSE for training data
train_prediction = problem2(w0, X_train)
train_linear_RSME = RMSE(train_prediction, y_train)
print("Training RMSE for linear regression using gradient descent is", train_linear_RSME)

# make predictions and calculate RMSE for test data
test_prediction = problem2(w0, X_test)

test_linear_RMSE = RMSE(test_prediction, y_test)
print("Testing RMSE for linear regression using gradient descent is", test_linear_RMSE)

Training RMSE for linear regression using gradient descent is 0.1277950313454118
Testing RMSE for linear regression using gradient descent is 0.14559783463038134


**Comparing Error Rates**

**Lab 3**
- Training: 0.12768967421762184
- Test: 0.12768967421762184

**Lab 4**
- Training: 0.1277950313454118
- Test: 0.14559783463038134

In [33]:
# compute loss function for ridge regression
def ridge_gradient_descent(x, y, w, lambda_value):
    left = (y - np.dot(x, w)).T
    right = (y - np.dot(x, w))
    min_W = np.dot(left, right) + (optimal_lambda * np.dot(w0.T, w0))
    return min_W

alpha = 5*(10**-5)

In [34]:
# takes in an NxK array and returns predictions as Nx1 array
def problem3(w, validation):
    predictions = []
    for new_X in validation:
        predictions.append(np.dot(w.T, new_X))
    return predictions

# finding the optimal lambda
lambda_value = 400
RMSEs = np.zeros((10,5))
minimum = 0

# do k-fold cross validation with k = 5
for segment in range(len(RMSEs)):
    for k in range(5):
        # train with 4/5 of our dataset
        X_train_k = np.concatenate((X_train[0:int(k*(len(X_train)/5))], X_train[int((k+1)*(len(X_train)/5)):int(len(X_train))]))
        y_train_k = np.concatenate((y_train[0:int(k*(len(y_train)/5))], y_train[int((k+1)*(len(y_train)/5)):int(len(y_train))]))
        
        # use remaining 1/5 dataset to validate results
        X_validate_k = X_train[int(k*(len(X_train)/5)):int((k+1)*(len(X_train)/5))]
        y_validate_k = y_train[int(k*(len(y_train)/5)):int((k+1)*(len(y_train)/5))]
        
        # perform ridge regression using gradient descent on each fold and each lambda
        L1 = 1
        L0 = 0
        
        # generate a Gaussian distribution for initial value based off how many features we have
        w0 = np.random.normal(0, 1, (X_train.shape[1],1))
        while (abs(L1 - L0) > (10**-5)):
            w1 = w0 - (alpha * (np.dot(X_train_k.T, (np.dot(X_train_k, w0) - y_train_k)) + (lambda_value * w0)))
            L0 = ridge_gradient_descent(X_train_k, y_train_k, w0, lambda_value)
            L1 = ridge_gradient_descent(X_train_k, y_train_k, w1, lambda_value)
            w0 = w1
        validated_results = []
        validated_results = problem3(w0, X_validate_k)
        # for each lambda, put all 5 of the different RMSEs on the same row
        RMSEs[segment, k] = RMSE(validated_results, y_validate_k) 
    lambda_value /= 2

print("RMSE values:\n",RMSEs, "\n")
    
# get average RMSE for each lambda and choose optimal lambda
for i in range(len(RMSEs)-1):
    if (sum(RMSEs[i+1]) / 5) < (sum(RMSEs[minimum]) / 5):
        minimum = i+1

for j in range(len(RMSEs)):
    print("𝜆 =", (400 / (2 ** j)), ":", sum(RMSEs[j]) / 5)

optimal_lambda = 400 / (2 ** minimum)

print("\nThe optimal 𝜆 is:", optimal_lambda)
print("Training RMSE for ridge regression using gradient descent is", (sum(RMSEs[minimum]) / 5))

RMSE values:
 [[0.16031764 0.14529638 0.15784575 0.14649755 0.13770169]
 [0.15049631 0.13910946 0.14500678 0.13852478 0.12994672]
 [0.14641962 0.13759955 0.13916373 0.13586991 0.12681816]
 [0.14521579 0.13738674 0.13641804 0.13500564 0.12687195]
 [0.14487902 0.13764173 0.13474423 0.13356439 0.12775887]
 [0.14538247 0.13771015 0.13361137 0.13451657 0.12888735]
 [0.14609972 0.13818297 0.13306128 0.13377324 0.12977529]
 [0.14643816 0.13878245 0.13310104 0.13458828 0.13067933]
 [0.14719936 0.13919435 0.13293082 0.13454885 0.13095068]
 [0.14655885 0.14127764 0.13307229 0.13470611 0.13142194]] 

𝜆 = 400.0 : 0.1495318037775622
𝜆 = 200.0 : 0.1406168077172081
𝜆 = 100.0 : 0.13717419293232766
𝜆 = 50.0 : 0.13617963386696483
𝜆 = 25.0 : 0.13571764742208164
𝜆 = 12.5 : 0.13602158279232798
𝜆 = 6.25 : 0.13617849922126912
𝜆 = 3.125 : 0.13671785454315827
𝜆 = 1.5625 : 0.1369648142055139
𝜆 = 0.78125 : 0.13740736498512002

The optimal 𝜆 is: 25.0
Training RMSE for ridge regression using gradient descent is 0.

In [35]:
# recompute w0 with optimal lambda and all data from training subset so we get results for our testing subset
L1 = 1
L0 = 0
# generate a Gaussian distribution for initial value based off how many features we have
w0 = np.random.normal(0, 1, (X_train.shape[1],1))
while (abs(L1 - L0) > (10**-5)):
    w1 = w0 - (alpha * (np.dot(X_train.T, (np.dot(X_train, w0) - y_train)) + (optimal_lambda * w0)))
    L0 = ridge_gradient_descent(X_train, y_train, w0, lambda_value)
    L1 = ridge_gradient_descent(X_train, y_train, w1, lambda_value)
    w0 = w1
test_predictions = []
test_predictions = problem3(w0, X_test)

test_ridge_RMSE = RMSE(test_predictions, y_test)
print("Testing RMSE for ridge regression using gradient descent is", test_ridge_RMSE)

Testing RMSE for ridge regression using gradient descent is 0.14573092500567844


**Problem 3 RMSE Values**
- Training: 0.13571764742208164
- Test: 0.14573092500567844